logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java

```diff
@@ -64,10 +63,12 @@ public Checkpoint write(String fileName, int pageNum, int firstUnackedPageNum, l
 
     @Override
     public void write(String fileName, Checkpoint checkpoint) throws IOException {
-        Path path = Paths.get(dirPath, fileName);
-        final byte[] buffer = new byte[BUFFER_SIZE];
         write(checkpoint, buffer);
-        Files.write(path, buffer, WRITE_OPTIONS);
+        buffer.flip();
+        try (FileOutputStream out = new FileOutputStream(Paths.get(dirPath, fileName).toFile())) {
```

Comment:
I had a curiosity, so I looked at the FileOutputStream.getChannel().write() effect. It calls sun.nio.ch.IOUtil.write() which will create a ByteBuffer if the given buffer is not a DirectBuffer. Cool -- this looks like it indeed avoids creating extra byte array.

Given this constraint, I think we need a comment on the `allocateDirect` call (line 34) to indicate this necessary in order to achieve your zero-copy goal. A comment something like (on line 34) `// We use a DirectBuffer here in order to achieve a zero-copy checkpoint write`.

Additionally for safety (regarding future maintenance), add an `assert buffer.isDirect()` at the top of the `write` method?

I am OK with your proposed change to use FileOutputStream, and my above comments are aimed at helping us not accidentally break your zero-copy intent.